### ISSUE TITLE
Add option to not upcast attention to save VRAM

### ISSUE THREAD

**Issue by kaiokendev** [oobabooga/text-generation-webui#2987]
**Description**

Anon pointed out that the attention does not need to be upcast to fp32 https://desuarchive.org/g/thread/94326476/#q94329535 and it saves significant memory: 2 GB on 2048 sequence length, and 3 GB on 4000 length from what I tested. He thought it might hurt precision/performance but I found that it does not hurt perplexity in any perceivable way for GPTQ models:

LLaMA 7B 4-bit 128g
```
{'name': 'fp32', 'length': 512, 'ppl': 8.67022705078125},
{'name': 'fp16', 'length': 512, 'ppl': 8.67022705078125},
{'name': 'fp32', 'length': 1024, 'ppl': 7.136347770690918},
{'name': 'fp16', 'length': 1024, 'ppl': 7.136347770690918},
{'name': 'fp32', 'length': 2048, 'ppl': 6.261297225952148},
{'name': 'fp16', 'length': 2048, 'ppl': 6.261251926422119}
```

It turns out this is already in exllama (but I didn't hear anyone mention it):
https://github.com/turboderp/exllama/blob/master/model.py#L412

We might as well add it here for GPTQ since it saves significant amount of memory. I didn't add a PR because not sure how you want this added -- a new attention flag option? Here is the change:

from
```
attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query_states.dtype)
```

to
```
attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float16).to(query_states.dtype)
```

basically, only change the dtype so it does not upcast to fp32. I was thinking to either add it to [modules/llama_attn_hijack.py](https://github.com/oobabooga/text-generation-webui/blob/main/modules/llama_attn_hijack.py) as `--fast_attn` or make it automatically use it if the backend is GPTQ or just do both. I don't know how it affects non-GPTQ, maybe @turboderp you already tested it?

**Comment by turboderp** [comment]
Attention scores end up being a relatively small tensor on short sequences (`(past_len + seq_len) * seq_len`, or `past_len + 1` usually). For longer sequences ExLlama uses `scaled_dot_product_attention` anyway, which chooses between the Torch implementation of Flash Attention or xformers as it sees fit.

But yes, I didn't see any downside to skipping the upcast in testing, although at the time I wasn't testing 8k sequences or anything like that.

I never got as far as looking at the implementation of `nn.functional.softmax` but when there is this little difference in ppl (or none at all in some cases), I suspect that the function works on the precision of the input tensor and `dtype=torch.float32` only causes an upcast of the result of the FP16 softmax. In which case downcasting immediately afterwards is just a waste of time and memory.

**Comment by kaiokendev** [comment]
@turboderp For SDPA, are you sure it is doing it? I did not look at exllama's code that much but from my experience with the SDPA, you need to set a specific context management call to allow Flash attention or mem-efficient (tiled) attention to run, otherwise it will always default back to standard attention which is what is already in modeling_llama.py (it does this silently btw). Idk if you tried it, but I have always been using xformers directly with `xformers.ops.memory_efficient_attention` and 
```
bias = (
            None
            if attention_mask is None or attention_mask[0, 0, 0, 1] == 0
            else xformers.ops.fmha.attn_bias.LowerTriangularMask()
        )
```

For the softmax, it upcasts the input tensor before doing calculation so the resulting calculation have greater precision, and the downcast is just truncation op, so the result is not the same as if the input tensor was always in the lower precision form. Ok, I will try also with long 8K sequences, I think the ppl loss will still be mostly imperceptible, but did you ever try on, i.e. 8-bit model? Or not since exllama is only meant for 4-bit

**Comment by turboderp** [comment]
Torch documentation is famously bad, so I'm not 100% sure, but both Flash and xformers should be [enabled by default](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html):

>All implementations are enabled by default. Scaled dot product attention attempts to automatically select the most optimal implementation based on the inputs. 

I guess I'll do some more tests later to see if I've been deceived, maybe.

It would make sense for the softmax to apply the dtype before the computation, but then I'm really surprised that there's no difference in perplexity at all until a 2048 sequence length, and even then nothing worth mentioning. The function is defined as `exp(x)/sum(exp(x))` so it is summing values across quite a large range. Even if it ends up not mattering in practice, I still wouldn't expect the output to be _identical_ for FP16 and FP32.

**Comment by turboderp** [comment]
Okay, I tested it a little more and there is a difference, it's just tiny. The `dtype=float32` softmax result differs from the `dtype=float16` result by up to about 2e-4 in my tests, and it still rounds to the exact same FP16 attention weights in my (limited) tests. So probably just something to consider if a model starts performing poorly at very long context lengths, but I don't see any reason for the upcast otherwise.

**Comment by Ph0rk0z** [comment]
Would this affect cards like the P40 since FP16 calculations slow things down at least by half?

**Comment by oobabooga** [comment]
If it doesn't affect performance in a noticeable way while reducing VRAM usage, I think that it probably can be a default

**Comment by turboderp** [comment]
Like most FP16 optimizations it probably does kill performance on the P40, though.

**Comment by Ph0rk0z** [comment]
Then it's good to just make it an option. At this point there is not much point to use AutoGPTQ anyway, unless you can't run exllama or have a different model architecture that is unsupported. 

I have still not seen much gain from any of the attention mechanisms on either pascal or ampere cards when using old gptq or autogptq. I enabled them all and there was hardly any movement in memory size or speed. Perhaps the upcast was why?

**Comment by kaiokendev** [comment]
It's unfortunate it will not work with P40, because it even helps during training. It reduces memory usage even more than xformers. I will try to make a PR with optional flag

**Comment by oobabooga** [comment]
Reopening because this is relevant.
